auth0/resource_auth0_attack_protection.go

```diff
@@ -304,125 +345,118 @@ func flattenBreachedPasswordProtection(bpd *management.BreachedPasswordDetection
-			ipt.Stage.PreLogin.Rate = Int(d, "rate")
-		})
+			stateAllowList := d.Get("allowlist").([]interface{})
+			allowlist := make([]string, len(stateAllowList))
+			for index, value := range stateAllowList {
+				allowlist[index] = value.(string)
+			}
+			ipt.AllowList = &allowlist
+
+			_, ok := d.GetOk("pre_login")
+			if ok {
+				if ipt.Stage == nil {
+					ipt.Stage = &management.Stage{}
+				}
+				if ipt.Stage.PreLogin == nil {
+					ipt.Stage.PreLogin = &management.PreLogin{}
+				}
+
+				List(d, "pre_login", IsNewResource(), HasChange()).Elem(func(d ResourceData) {
+					ipt.Stage.PreLogin.MaxAttempts = Int(d, "max_attempts")
+					ipt.Stage.PreLogin.Rate = Int(d, "rate")
+				})
+			}
 
-		List(d, "pre_user_registration").Elem(func(d ResourceData) {
-			ipt.Stage.PreUserRegistration.MaxAttempts = Int(d, "max_attempts")
-			ipt.Stage.PreUserRegistration.Rate = Int(d, "rate")
+			_, ok = d.GetOk("pre_user_registration")
+			if ok {
+				if ipt.Stage == nil {
+					ipt.Stage = &management.Stage{}
+				}
+				if ipt.Stage.PreUserRegistration == nil {
+					ipt.Stage.PreUserRegistration = &management.PreUserRegistration{}
+				}
+
+				List(d, "pre_user_registration", IsNewResource(), HasChange()).Elem(func(d ResourceData) {
+					ipt.Stage.PreUserRegistration.MaxAttempts = Int(d, "max_attempts")
+					ipt.Stage.PreUserRegistration.Rate = Int(d, "rate")
+				})
+			}
 		})
-	})
+	}
 
-	return ipt
+	return &ipt
 }
 
 func expandBruteForceProtection(d *schema.ResourceData) *management.BruteForceProtection {
-	bfp := &management.BruteForceProtection{}
+	var bfp management.BruteForceProtection
 
-	List(d, "brute_force_protection", IsNewResource(), HasChange()).Elem(func(d ResourceData) {
-		shields := []string{}
-		for _, s := range d.Get("shields").([]interface{}) {
-			shields = append(shields, fmt.Sprintf("%s", s))
-		}
+	_, ok := d.GetOk("brute_force_protection")
+	if ok {
+		List(d, "brute_force_protection", IsNewResource(), HasChange()).Elem(func(d ResourceData) {
+			bfp.Enabled = Bool(d, "enabled")
 
-		allowlist := []string{}
-		for _, a := range d.Get("allowlist").([]interface{}) {
-			allowlist = append(allowlist, fmt.Sprintf("%s", a))
-		}
+			stateShields := d.Get("shields").([]interface{})
+			shields := make([]string, len(stateShields))
+			for index, value := range stateShields {
+				shields[index] = value.(string)
+			}
+			bfp.Shields = &shields
 
-		bfp = &management.BruteForceProtection{
-			Enabled:     Bool(d, "enabled"),
-			Shields:     &shields,
-			AllowList:   &allowlist,
-			Mode:        String(d, "mode"),
-			MaxAttempts: Int(d, "max_attempts"),
-		}
-	})
+			stateAllowList := d.Get("allowlist").([]interface{})
+			allowlist := make([]string, len(stateAllowList))
+			for index, value := range stateAllowList {
+				allowlist[index] = value.(string)
+			}
+			bfp.AllowList = &allowlist
+
+			bfp.Mode = String(d, "mode")
+			bfp.MaxAttempts = Int(d, "max_attempts")
+		})
+	}
 
-	return bfp
+	return &bfp
```

Comment:
IMO, this is a huge downgrade in our code quality and it appears to be driven by fear that we don't actually know what the API will return. That's a question that we _should_ be able to answer before we attempt to implement any fix here.

Reply:
I would like to understand better what do you mean by downgrading our code quality? What in your eyes is a downgrade?

We do know what the API will return. The problem is that the `stage` property ends up being omitted in certain cases. You can check the suspicious-ip-throttling handler in APIv2.

```
// Do not return `stage` attribute with custom thresholds if....
```